src/glue.c

```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
   return ret;
 }
 
+void maybe_decrement_ref(PyObject **obj) {
+  if (*obj) {
+    Py_DECREF(obj);
+  }
+}
+
+#define SCOPED_CLEANUP __attribute__((cleanup(maybe_decrement_ref)))
+#define IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(ptr, value) \
+  if (!(ptr)) {                                           \
+    PyErr_Print();                                        \
+    return (value);                                       \
+  }
+
+PyObject *make_py_date(int y, int m, int d) {
+  PyDateTime_IMPORT;
+  PyObject *date = PyDate_FromDate(y, m, d);
+  if (!date) {
+    PyErr_Print();
+  }
+  return date;
+}
+
+PyObject *make_py_time(int h, int min, int sec, int mi) {
+  PyDateTime_IMPORT;
+  PyObject *time = PyTime_FromTime(h, min, sec, mi);
+  if (!time) {
+    PyErr_Print();
+  }
+  return time;
+}
+
+PyObject *make_py_datetime(int y, int m, int d, int h, int min, int sec,
+                           int mi) {
+  PyDateTime_IMPORT;
+  PyObject *datetime = PyDateTime_FromDateAndTime(y, m, d, h, min, sec, mi);
+  if (!datetime) {
+    PyErr_Print();
+  }
+  return datetime;
+}
+
+PyObject *make_py_delta(int days, int sec, int micros) {
+  PyDateTime_IMPORT;
+  PyObject *delta = PyDelta_FromDSU(days, sec, micros);
+  if (!delta) {
+    PyErr_Print();
+  }
+  return delta;
+}
+
+PyObject *mg_date_to_py_date(const mg_date *date) {
+  PyDateTime_IMPORT;
+  SCOPED_CLEANUP PyObject *unix_epoch = make_py_date(1970, 1, 1);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(unix_epoch, NULL)
+  SCOPED_CLEANUP PyObject *date_as_delta =
+      make_py_delta(mg_date_days(date), 0, 0);
+  SCOPED_CLEANUP PyObject *method_name = PyUnicode_FromString("__add__");
+  PyObject *result_date =
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, date_as_delta, NULL);
+  if (!result_date) {
+    PyErr_Print();
+  }
+  return result_date;
+}
+
+PyObject *mg_local_time_to_py_time(const mg_local_time *lt) {
+  PyDateTime_IMPORT;
+  SCOPED_CLEANUP PyObject *unix_epoch =
+      make_py_datetime(1970, 1, 1, 0, 0, 0, 0);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(unix_epoch, NULL)
+  SCOPED_CLEANUP PyObject *numerator =
+      PyLong_FromLong(mg_local_time_nanoseconds(lt));
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(numerator, NULL)
+  SCOPED_CLEANUP PyObject *denominator = PyLong_FromLong(1000000000);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(denominator, NULL)
+  SCOPED_CLEANUP PyObject *div_result =
+      PyNumber_TrueDivide(numerator, denominator);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(div_result, NULL)
+  SCOPED_CLEANUP PyObject *method_name =
+      PyUnicode_FromString("utcfromtimestamp");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(method_name, NULL)
+  SCOPED_CLEANUP PyObject *result =
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, div_result, NULL);
```

Comment:
I think you are calling a [class method](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp) on a concrete instance of a `datetime` object. I think you should get `datetime.datetime` type as a `PyObject` and call the `utcfromtimestamp` on that PyObject and not on a concrete instance of `datetime.datetime`.

I saw this multiple time, but I won't comment on all of them.

Reply:
I have been trying to figure this out but I don't know. It seems like, that if you convert `PyObjectType` to a `PyObject` you are actually creating an instance of its class. I could (potentially) reach the static method somehow through its [structure](https://github.com/python/cpython/blob/main/Include/datetime.h#L200) but this seems extremely convoluted. On the other hand, you can always call a noninstance method from an object of this class without an issue which is basically what is being done in the code above with the `unix_epoch`. Moreover, `CallMethod*` requires `PyObject` which implies an instance of an object (because a PyObject is just a wrapper for a PyObjectType + a reference count). I will investigate more and try to figure a way to do this. Maybe you have an idea?

Reply:
Could you use `PyDateTime_FromTimestamp` and `PyDateTime_TimeZone_UTC`?

Reply:
@antonio2368 I could use  `PyDateTime_FromTimestamp`. However, I have to convert  the units to microseconds +  allocate the `PyObject` which is a little bit cumbersome.